source/fuzz/transformation_replace_copy_memory_with_load_store.cpp

```diff
@@ -0,0 +1,118 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_replace_copy_memory_with_load_store.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationReplaceCopyMemoryWithLoadStore::
+    TransformationReplaceCopyMemoryWithLoadStore(
+        const spvtools::fuzz::protobufs::
+            TransformationReplaceCopyMemoryWithLoadStore& message)
+    : message_(message) {}
+
+TransformationReplaceCopyMemoryWithLoadStore::
+    TransformationReplaceCopyMemoryWithLoadStore(
+        uint32_t source_value,
+        const protobufs::InstructionDescriptor& instruction_descriptor) {
+  message_.set_source_value(source_value);
+  *message_.mutable_instruction_descriptor() = instruction_descriptor;
+}
+
+bool TransformationReplaceCopyMemoryWithLoadStore::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |message_.source_value| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.source_value())) {
+    return false;
+  }
+  // The instruction to be replaced must be defined and of opcode
+  // OpCopyMemory.
+  auto copy_memory_instruction =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!copy_memory_instruction ||
+      copy_memory_instruction->opcode() != SpvOpCopyMemory) {
+    return false;
+  }
+  // It must be valid to insert the OpStore and OpLoad instruction before it.
+  return (fuzzerutil::CanInsertOpcodeBeforeInstruction(
+              SpvOpStore, copy_memory_instruction) &&
+          fuzzerutil::CanInsertOpcodeBeforeInstruction(
+              SpvOpLoad, copy_memory_instruction));
+}
+
+void TransformationReplaceCopyMemoryWithLoadStore::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto copy_memory_instruction =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  // |copy_memory_instruction| must be defined.
+  assert(copy_memory_instruction &&
+         copy_memory_instruction->opcode() == SpvOpCopyMemory &&
+         "The required OpCopyMemory instruction must be defined.");
+
+  // Get types of ids used as a source and target of |copy_memory_instruction|.
+  auto target = ir_context->get_def_use_mgr()->GetDef(
+      copy_memory_instruction->GetSingleWordInOperand(0));
+  auto source = ir_context->get_def_use_mgr()->GetDef(
+      copy_memory_instruction->GetSingleWordInOperand(1));
+  auto target_type_opcode =
+      ir_context->get_def_use_mgr()->GetDef(target->type_id())->opcode();
+  auto source_type_opcode =
+      ir_context->get_def_use_mgr()->GetDef(source->type_id())->opcode();
+
+  // Both operands must be pointers.
+  assert(target_type_opcode == SpvOpTypePointer &&
+         source_type_opcode == SpvOpTypePointer &&
+         "Operands must be of type OpTypePointer");
```

Comment:
All this is merely a coherence check.

I like coherence checks in general, so happy to keep this, but please indicate with a comment that it's a coherence check - because it introduces various temporary variables a reader of the code might at first think they were going to be used in the functionality of Apply.

Also, you may need to do:

(void)temporary_name; // Keep release-mode compilers happy

to avoid compiler warnings - in release mode the assertions don't exist so you might get messages about unused variables.